File: app/src/main/java/app/we/go/emojidraw/widget/DrawingView.java
package app.we.go.emojidraw.widget;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.support.annotation.ColorInt;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

import app.we.go.emojidraw.R;
import app.we.go.emojidraw.model.Stroke;
import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;

/**
 * View that allows the user to draw into it. Offers some configuration options:
 * background color, paint color and stroke width (as well as, obviously, size).
 * Allows for undo and clear option. Exposes an observable with the strokes currently drawn.
 */
public class DrawingView extends View {

    private Paint paint;
    private Path currentPath;
    private ArrayList<Path> paths;
    private Context context;
    private Canvas drawingCanvas;
    private Bitmap drawingBitmap;

    private List<Stroke> strokes;
    private Stroke currentStroke;

    private final PublishSubject<List<Stroke>> pubsub = PublishSubject.create();
    private @ColorInt int backgroundColor;
    private @ColorInt int paintColor;
    private float strokeWidth;

    public DrawingView(Context c) {
        super(c);
        init(context, null);
    }

    public DrawingView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    public DrawingView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    @Override
    protected void onSizeChanged(final int w, final int h, final int oldw, final int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        drawingBitmap = Bitmap.createBitmap(getWidth(), getHeight(), Bitmap.Config.RGB_565);
        drawingCanvas = new Canvas(drawingBitmap);
        clearDrawingCanvas();
    }

    private void init(Context context, @Nullable AttributeSet attrs) {
        this.context=context;

        TypedArray a = context.getTheme().obtainStyledAttributes(
                attrs,
                R.styleable.DrawingView,
                0, 0);

        try {
            backgroundColor = a.getColor(R.styleable.DrawingView_dv_backgroundColor,
                    ContextCompat.getColor(context, android.R.color.white));
            paintColor = a.getColor(R.styleable.DrawingView_dv_paintColor,
                    ContextCompat.getColor(context, android.R.color.black));
            strokeWidth = a.getFloat(R.styleable.DrawingView_dv_strokeWidth, 10f);
        } finally {
            a.recycle();
        }

        paint = new Paint();
        paint.setAntiAlias(true);
        paint.setDither(true);
        paint.setColor(paintColor);
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeJoin(Paint.Join.ROUND);
        paint.setStrokeCap(Paint.Cap.ROUND);
        paint.setStrokeWidth(strokeWidth);

        currentPath = new Path();
        paths = new ArrayList<>();

        currentStroke = new Stroke();
        strokes = new ArrayList<>();
    }


    @Override
    protected void onDraw(Canvas canvas) {
        for (int i = 0; i < paths.size(); i++) {
            drawingCanvas.drawPath(paths.get(i), paint);
        }

        drawingCanvas.drawPath(currentPath, paint);

        canvas.drawBitmap(drawingBitmap, 0, 0, paint);
    }


    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouchEvent(MotionEvent event) {

        switch (event.getAction()){

            case MotionEvent.ACTION_DOWN:
                // Little hack to make sure "points" can be drawn
                currentPath.moveTo(event.getX()-1, event.getY()-1);
                currentPath.lineTo(event.getX()+1, event.getY()+1);
                currentStroke.addPoint((int) event.getX(), (int) event.getY());
                break;

            case MotionEvent.ACTION_MOVE:
                currentPath.lineTo(event.getX(), event.getY());
                currentStroke.addPoint((int) event.getX(), (int) event.getY());
                invalidate();
                break;

            case MotionEvent.ACTION_UP:
                paths.add(currentPath);
                strokes.add(currentStroke);
                currentPath = new Path();
                currentStroke = new Stroke();
                invalidate(); // need this here as well for "points" that are up and down events (without passing from move)
                onStrokeAdded();
                break;
        }

        return true;
    }

    private void onStrokeAdded() {
       pubsub.onNext(strokes);
    }

    /**
     * An Observable that emits, on each stroke (or undo/clear actions) the list strokes that are
     * currently drawn on the canvas.
     * @return
     */
    public Observable<List<Stroke>> getStrokesObservable() {
        return pubsub;
    }

    /**
     * Clear the view completely
     */
    public void clear() {
        paths.clear();
        strokes.clear();
        pubsub.onNext(strokes);
        clearDrawingCanvas();
        invalidate();
    }

    /**
     * Undo the last stroke that was drawn.
     */
    public void undo() {
        if (!paths.isEmpty()) {
            clearDrawingCanvas();
            paths.remove(paths.size() - 1);
            strokes.remove(strokes.size() - 1);
            pubsub.onNext(strokes);
            invalidate();
        }
    }

    private void clearDrawingCanvas() {
        drawingCanvas.drawColor(backgroundColor);
    }


}

File: app/src/main/java/app/we/go/emojidraw/widget/DrawingView.kt
package app.we.go.emojidraw.widget

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import androidx.annotation.ColorInt
import androidx.core.content.ContextCompat
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View

import app.we.go.emojidraw.R
import app.we.go.emojidraw.model.Stroke
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject

/**
 * View that allows the user to draw into it. Offers some configuration options:
 * background color, paint color and stroke width (as well as, obviously, size).
 * Allows for undo and clear option. Exposes an observable with the strokes currently drawn.
 */
class DrawingView @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0, defStyleRes: Int = 0
) : View(context, attrs, defStyleAttr, defStyleRes) {

    private val paint: Paint
    private var currentPath: Path = Path()
    private val paths: MutableList<Path> = mutableListOf()
    private lateinit var drawingCanvas: Canvas
    private lateinit var drawingBitmap: Bitmap

    private val strokes: MutableList<Stroke> = mutableListOf()
    private var currentStroke: Stroke = Stroke()

    private val pubsub = PublishSubject.create<List<Stroke>>()

    @ColorInt
    private val bgColor: Int
    @ColorInt
    private val paintColor: Int

    private val strokeWidth: Float


    init {
        val typedArray = context.obtainStyledAttributes(
            attrs,
            R.styleable.DrawingView,
            0, 0)

        bgColor = typedArray.getColor(R.styleable.DrawingView_dv_backgroundColor,
            ContextCompat.getColor(context, android.R.color.white))
        paintColor = typedArray.getColor(R.styleable.DrawingView_dv_paintColor,
            ContextCompat.getColor(context, android.R.color.black))
        strokeWidth = typedArray.getFloat(R.styleable.DrawingView_dv_strokeWidth, 10f)

        typedArray.recycle()

        paint = Paint().apply {
            isAntiAlias = true
            isDither = true
            color = paintColor
            style = Paint.Style.STROKE
            strokeJoin = Paint.Join.ROUND
            strokeCap = Paint.Cap.ROUND
            strokeWidth = this@DrawingView.strokeWidth
        }
    }


    /**
     * An Observable that emits, on each stroke (or undo/clear actions) the list strokes that are
     * currently drawn on the canvas.
     * @return
     */
    val strokesObservable: Observable<List<Stroke>>
        get() = pubsub


    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        drawingBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565)
        drawingCanvas = Canvas(drawingBitmap)
        clearDrawingCanvas()
    }


    override fun onDraw(canvas: Canvas) {
        (paths + currentPath).map {
            drawingCanvas.drawPath(it, paint)
        }

        canvas.drawBitmap(drawingBitmap, 0f, 0f, paint)
    }


    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {

        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                // Little hack to make sure "points" can be drawn
                currentPath.moveTo(event.x - 1, event.y - 1)
                currentPath.lineTo(event.x + 1, event.y + 1)
                currentStroke.addPoint(event.x.toInt(), event.y.toInt())
            }

            MotionEvent.ACTION_MOVE -> {
                currentPath.lineTo(event.x, event.y)
                currentStroke.addPoint(event.x.toInt(), event.y.toInt())
                invalidate()
            }

            MotionEvent.ACTION_UP -> {
                paths.add(currentPath)
                strokes.add(currentStroke)
                currentPath = Path()
                currentStroke = Stroke()
                invalidate() // need this here as well for "points" that are up and down events (without passing from move)
                onStrokeAdded()
            }
        }

        return true
    }

    private fun onStrokeAdded() {
        pubsub.onNext(strokes)
    }

    /**
     * Clear the view completely
     */
    fun clear() {
        paths.clear()
        strokes.clear()
        pubsub.onNext(strokes)
        clearDrawingCanvas()
        invalidate()
    }

    /**
     * Undo the last stroke that was drawn.
     */
    fun undo() {
        if (!paths.isEmpty()) {
            clearDrawingCanvas()
            paths.removeAt(paths.size - 1)
            strokes.removeAt(strokes.size - 1)
            pubsub.onNext(strokes)
            invalidate()
        }
    }

    private fun clearDrawingCanvas() {
        drawingCanvas.drawColor(bgColor)
    }

}